### ISSUE TITLE
Per-network bitcoin.conf files

### ISSUE THREAD

**Issue by rebroad** [bitcoin/bitcoin#9374]
When running with the -testnet option, I would expect the bitcoin.conf file in the testnet3 directory to be used, yet it uses the mainnet config file instead, yet the debug.log indicates the datadir is the testnet3 dir:-

```
2016-12-19 03:08:58.629380 Using data directory /home/rebroad/.bitcoin/testnet3
2016-12-19 03:08:58.629453 Using config file /home/rebroad/.bitcoin/bitcoin.conf
```

**Comment by laanwj** [comment]
There has never been any functionality to read per-network config files. Config file is always read from the root of the data directory.

However this is more complicated to implement than you'd think, not only because of backward compatiblity, but because bitcoin.conf in the data directory can specify the network to use e.g. `testnet=1` or `regtest=1`. This is used by the tests. It is also useful in the case of specifying a .conf file with `-conf` that is outside any data directory (used for system-wide setups).

**Comment by laanwj** [comment]
One idea of doing this is to keep `bitcoin.conf` as global configuration file, then after that is loaded (and thus the network determined) load an optional per-network configuration that can override these parameters, e.g.:

- `${DATADIR}/network.conf`
- `${DATADIR}/testnet3/network.conf`
- `${DATADIR}/regtest/network.conf`

**Comment by rebroad** [comment]
@laanwj Hmmm. sounds messy; for example, how would 'addnode='s be overridden if there were none mentioned in the testnet config file? I often want to add nodes for the mainnet configuration but not for the testnet configuration. For backwards compatibility (which I agree is important), I'd suggest that if the conf file for testnet or regtest does not exist then it defaults back to the mainnet config file.

**Comment by laanwj** [comment]
My proposal above allows for full flexibility here.

Bitcoin Core will look, in a second configuration stage, for a `network.conf` in the per-network data directory and process this.

If you wanted to add testnet-specific addnodes you'd put these in `${DATADIR}/testnet3/network.conf`. If you want to add mainnet-specific addnotes you'd put these in `${DATADIR}/network.conf`. In the "global" bitcoin.conf you'd not put any addnodes, you can put configuration there that you want to always apply.

**Comment by morcos** [comment]
@laanwj +1 for network.conf idea

**Comment by rebroad** [comment]
I suspect, I would just prefer to use one .conf file per network, so ideally this would allow me to abandon the bitcoin.conf file, and just use network.conf files.

**Comment by jonasschnelli** [comment]
+1 for @laanwj network.conf idea.

@rebroad: your solution would no longer be downward compatible (by dropping support for the network selection in bitcoin.conf).

**Comment by rebroad** [comment]
@jonasschnelli I wasn't proposing to drop support for anything in the bitcoin.conf file. My solution sounds the same as @laanwj's except that I would have kept the conf file as bitcoin.conf and not created a new config file.

**Comment by jonasschnelli** [comment]
@rebroad: but how would it be practical to switch the network (`regtest=1`, `testnet=1`) if you remove the general `bitcoin.conf` file?

**Comment by btcdrak** [comment]
+1 for @laanwj network.conf idea.

**Comment by laanwj** [comment]
> I suspect, I would just prefer to use one .conf file per network, so ideally this would allow me to abandon the bitcoin.conf file, and just use network.conf files.

Yes, network.conf file should be looked for irrespective of whether bitcoin.conf exists (that's also easiest to implement).

> @rebroad: but how would it be practical to switch the network (regtest=1, testnet=1) if you remove the general bitcoin.conf file?

He chooses to not use that functionality - but for any solution here it should still be supported - those are orthogonal things :)

**Comment by gmaxwell** [comment]
I often intentionally use the same config files on mainnet and testnet.  The two stage thing would be dandy.

**Comment by rebroad** [comment]
I'm going to try and raise an issue with github regarding these message above (they are produced whenever I re-push a branch containing this commit).

**Comment by rebroad** [comment]
rebased following #9243

**Comment by jnewbery** [comment]
Closed by #11862 and #10267. It is now possible to create network-specific sections or entire config files.
